### ISSUE TITLE
ssl: misleading `FileNotFoundError: [Errno 2] No such file or directory` on `do_handshake`

### ISSUE THREAD

**Issue by u1735067** [python/cpython#102547]
# Bug report

If a ssl handshake is interrupted (due to a buggy server for example), the exception `FileNotFoundError: [Errno 2] No such file or directory` is raised, which might be misleading. This issue was previously raised in https://github.com/python/cpython/issues/79505 / https://bugs.python.org/issue35324, I'm not sure this is exactly the same case however.

This was tested with Python 3.7.5 and 3.10.0. Also under 3.7.5, the `client2` test returns `OSError: [Errno 0] Error` which isn't explicit either, but this seems fixed in 3.10.0 with the exception `ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:997)`, which is probably what should be raised for `client` too?

To reproduce, use the following file to start a server then a client in another console:
```sh
py poc.py server
py poc.py client
```

<details><summary><code>poc.py</code></summary>
<p>

```py
import sys
import ssl  # https://docs.python.org/3/library/ssl.html
import socket  # https://docs.python.org/3/library/socket.html


intf = '127.0.0.1'
port = 6789


def client(intf, port, err=1):
    """Simple TLS client"""
    context = ssl.create_default_context()
    context.check_hostname = False
    # context.verify_mode = ssl.CERT_NONE
    # context.post_handshake_auth = False

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.settimeout(2)
        sock.connect((intf, port))
        if err == 1:
            # import requests;
            # requests.get(f'https://{intf}:{port}/', verify=False)
            # OR
            # import urllib3;
            # urllib3.connection_from_url(f'https://{intf}:{port}/').request('GET', '/')
            context.wrap_socket(sock)  # FileNotFoundError: [Errno 2] No such file or directory
        else:
            # import urllib.request, ssl;
            # ctx = ssl.create_default_context();
            # ctx.check_hostname = False;
            # urllib.request.urlopen(f'https://{intf}:{port}/', context=ctx)
            context.wrap_socket(sock, server_hostname=intf)  # OSError: [Errno 0] Error

    # https://bugs.python.org/issue35324
    r"""
    Traceback (most recent call last):
    File "poc.py", line .., in <module>
        context.wrap_socket(sock)
    File "C:\Apps\Python375\lib\ssl.py", line 423, in wrap_socket
        session=session
    File "C:\Apps\Python375\lib\ssl.py", line 870, in _create
        self.do_handshake()
    File "C:\Apps\Python375\lib\ssl.py", line 1139, in do_handshake
        self._sslobj.do_handshake()
    FileNotFoundError: [Errno 2] No such file or directory
    """


def server(intf, port):
    """Buggy server that close socket during handshake"""
    def sni_callback(
        sock: ssl.SSLSocket,
        req_hostname: str,
        cb_context: ssl.SSLContext
    ):
        sock.close()
        return None

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    s.bind((intf, port))
    s.listen(8)
    default_timeout = s.gettimeout()
    print("Serving on {}:{}".format(intf, port))

    try:
        while True:
            # https://stackoverflow.com/questions/34871191/cant-close-socket-on-keyboardinterrupt
            s.settimeout(0.5)
            try:
                conn, (host, port) = s.accept()
            except socket.timeout:
                continue
            s.settimeout(default_timeout)
            print('connected')

            context = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH)
            # https://stackoverflow.com/questions/41996833/using-ssl-context-set-servername-callback-in-python
            context.set_servername_callback(sni_callback)
            #context.load_cert_chain(certfile="cert.pem", keyfile="cert.key")
            try:
                ssl_sock = context.wrap_socket(conn, server_side=True)
            except OSError:
                pass

            conn.close()
            # try:
            #     handle_client(ssl_sock, (host, port))
            # finally:
            #     conn.close()

    except KeyboardInterrupt:
        s.close()


if __name__ == '__main__':
    if len(sys.argv) < 2 or sys.argv[1] not in {'server', 'client', 'client2'}:
        print('Usage: poc.py {server,client}')
        raise SystemExit()
    if sys.argv[1] == 'client':
        client(intf, port)
    if sys.argv[1] == 'client2':
        client(intf, port, err=2)
    if sys.argv[1] == 'server':
        server(intf, port)
```

</p>
</details>

# Your environment

- CPython versions tested on: 3.7.5, 3.10.0
- Operating system and architecture: Win10 x64 19045

**Comment by arhadthedev** [comment]
@jackjansen, @dstufft, @alex (as ssl [experts](https://devguide.python.org/core-developers/experts/))

**Comment by gagebenne** [comment]
Unable to replicate on macOS -- attempted with Python `3.11.3`, `3.10.11`, `3.12.0a7+`.

**Comment by Janturba** [comment]
Just my two cents... I recently had fun with this too. Kept getting FileNotFound. All I ended up doing was renamed the certificate to "certificate.pem" and the key to "private.key" and it started to work. Before this the files were named "hsm_subca_cert.pem" and "hsm_subca_key.pem". I did not move the files or changed perms. Just a rename. So perhaps some issue with the parser and the name having "_" in them.

**Comment by u1735067** [comment]
This seems fixed in latest Python 3.12.0rc1 (https://www.python.org/ftp/python/3.12.0/python-3.12.0rc1-embed-amd64.zip) :)

```
PS C:\Tmp\py12> .\python.exe .\poc.py client
Traceback (most recent call last):
  File "C:\Tmp\py12\poc.py", line 101, in <module>
    client(intf, port)
  File "C:\Tmp\py12\poc.py", line 26, in client
    context.wrap_socket(sock)  # FileNotFoundError: [Errno 2] No such file or directory
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ssl.py", line 455, in wrap_socket
  File "ssl.py", line 1013, in _create
  File "ssl.py", line 1284, in do_handshake
ssl.SSLEOFError: [SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1000)
```

```
PS C:\Tmp\py12> .\python.exe .\poc.py client2
Traceback (most recent call last):
  File "C:\Tmp\py12\poc.py", line 103, in <module>
    client(intf, port, err=2)
  File "C:\Tmp\py12\poc.py", line 32, in client
    context.wrap_socket(sock, server_hostname=intf)  # OSError: [Errno 0] Error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ssl.py", line 455, in wrap_socket
  File "ssl.py", line 1013, in _create
  File "ssl.py", line 1284, in do_handshake
ssl.SSLEOFError: [SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1000)
```

It was not in Python 3.11.4 (https://www.python.org/ftp/python/3.11.4/python-3.11.4-embed-amd64.zip):

```
PS C:\Tmp\py11> .\python.exe .\poc.py client
Traceback (most recent call last):
  File "C:\Tmp\py11\poc.py", line 101, in <module>
    client(intf, port)
  File "C:\Tmp\py11\poc.py", line 26, in client
    context.wrap_socket(sock)  # FileNotFoundError: [Errno 2] No such file or directory
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ssl.py", line 517, in wrap_socket
  File "ssl.py", line 1075, in _create
  File "ssl.py", line 1346, in do_handshake
FileNotFoundError: [Errno 2] No such file or directory
```

```
PS C:\Tmp\py11> .\python.exe .\poc.py client2
Traceback (most recent call last):
  File "C:\Tmp\py11\poc.py", line 103, in <module>
    client(intf, port, err=2)
  File "C:\Tmp\py11\poc.py", line 32, in client
    context.wrap_socket(sock, server_hostname=intf)  # OSError: [Errno 0] Error
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ssl.py", line 517, in wrap_socket
  File "ssl.py", line 1075, in _create
  File "ssl.py", line 1346, in do_handshake
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1002)
```

I guess it is related to the OpenSSL update (1.1.1u -> 3.0.9) ; ~btw I'm not sure where to find the release OpenSSL bundled version in docs or source code, except downloading the release and inspecting the libssl dll?~ -> nevermind, found https://docs.python.org/3/whatsnew/changelog.html#windows.
